odl/discr/discr_ops.py

```diff
@@ -151,6 +159,226 @@ def adjoint(self):
+        discr_kwargs = kwargs.pop('discr_kwargs', {})
+
+        if range is None:
+            if ran_shp is None:
+                raise ValueError('either `range` or `ran_shp` must be '
+                                 'given')
+
+            num_left = normalized_scalar_param_list(
+                num_left, domain.ndim, param_conv=safe_int_conv)
+            self.__num_left = tuple(num_left)
+
+            range = _resize_discr(domain, ran_shp, num_left, discr_kwargs)
+
+        elif ran_shp is None:
+            if num_left is not None:
+                raise ValueError('`num_left` can only be combined with '
+                                 '`ran_shp`')
+
+            if not np.allclose(range.cell_sides, domain.cell_sides):
+                raise ValueError(
+                    'cell sides of domain and range differ significantly '
+                    '(difference {})'
+                    ''.format(range.cell_sides - domain.cell_sides))
+
+            self.__num_left = self._num_left_from_spaces(domain, range)
+
+        else:
+            raise ValueError('cannot combine `range` with `ran_shape`')
+
+        pad_mode = kwargs.pop('pad_mode', 'constant')
+        self.__pad_mode = str(pad_mode).lower()
+        if self.pad_mode not in _SUPPORTED_PAD_MODES:
+            raise ValueError("`pad_mode` '{}' not understood".format(pad_mode))
+
+        self.__pad_const = float(kwargs.pop('pad_const', 0.0))
+
+        # padding mode 'constant' with `pad_const != 0` is not linear
+        linear = (self.pad_mode != 'constant' or self.pad_const == 0.0)
+
+        super().__init__(domain, range, linear=linear)
+
+    @property
+    def num_left(self):
+        """Number of cells added to/removed from the left."""
+        return self.__num_left
+
+    @property
+    def pad_mode(self):
+        """Padding mode used by this operator."""
+        return self.__pad_mode
+
+    @property
+    def pad_const(self):
+        """Constant used by this operator in case of constant padding."""
+        return self.__pad_const
+
+    def _call(self, x, out):
+        """Implement ``self(x, out)``."""
+        # TODO: simplify once context manager is available
+        out[:] = resize_array(x.asarray(), self.range.shape,
+                              num_left=self.num_left, pad_mode=self.pad_mode,
+                              pad_const=self.pad_const, out=out.asarray())
+
+    def derivative(self, point):
+        """Derivative of this operator.
+
+        For the particular case of constant padding with non-zero
+        constant, the derivative is the corresponding zero-padding
+        variant. In all other cases, this operator is linear, i.e.
+        the derivative is equal to ``self``.
+        """
+        if self.pad_mode == 'constant' and self.pad_const != 0:
+            return ResizingOperator(
+                domain=self.domain, range=self.range, pad_mode='constant',
+                pad_const=0.0)
```

Comment:
Seem to miss several arguments here, like `discr_kwargs` and, `offset` and `ran_shape`.